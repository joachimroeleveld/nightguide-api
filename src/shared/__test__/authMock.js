const { TEST_USER_1 } = require('../../shared/__test__/fixtures/userFixtures');
const auth = require('../../framework/auth');
const User = require('../../routes/users/userModel');

function mockUserAuth(sandbox = global.sandbox) {
  const user = new User(TEST_USER_1);
  user.checkRole = sandbox.stub().returns(true);
  sandbox.stub(auth, 'getUserFromToken').resolves(user);
}

function restoreUserAuth() {
  auth.getUserFromToken.restore();
}

function mockAppClientAuth(sandbox = global.sandbox) {
  sandbox.stub(auth, 'checkAppTokenHeader').returns(true);
}

function restoreAppClientAuth() {
  auth.checkAppTokenHeader.restore();
}

module.exports = {
  mockUserAuth,
  restoreUserAuth,
  mockAppClientAuth,
  restoreAppClientAuth,
};
