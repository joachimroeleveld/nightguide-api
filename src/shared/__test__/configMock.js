const config = require('../config');

const mockedEnv = {
  ...process.env,
};

function mockKey(key, value, sandbox = global.sandbox) {
  mockedEnv[key] = value;
  sandbox.stub(config, 'get').callsFake(key => mockedEnv[key]);
}

module.exports = {
  mockKey,
};
