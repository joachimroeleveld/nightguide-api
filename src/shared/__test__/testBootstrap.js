const dotenv = require('dotenv');
dotenv.load();

const sgMail = require('@sendgrid/mail');
const mongoose = require('mongoose');
const config = require('../config');

const sandbox = require('sinon').createSandbox();
const { mockAuth } = require('./authMock');
const configMock = require('./configMock');

const { createExpressApp } = require('../../framework/expressServer');

beforeAll(cb => {
  // Use global sandbox for mocks that you want to persist across multiple test suites
  if (!global.sandbox) {
    global.sandbox = sandbox;
  }

  mockAuth();

  configMock.mockKey('HOST', 'http://localhost:8080');

  // Mock SendGrid
  sandbox.stub(sgMail, 'send').resolves();

  global.app = createExpressApp();

  mongoose
    .connect(
      config.get('MONGO_URI'),
      { useNewUrlParser: true, dbName: 'main-test' }
    )
    .then(() => cb())
    .catch(err => cb(err));
});

afterAll(async () => {
  mongoose.connection.close();

  global.sandbox.restore();
});
