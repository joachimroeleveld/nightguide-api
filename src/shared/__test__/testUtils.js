const mongoose = require('mongoose');
const _ = require('lodash');

const { deepFreeze } = require('../util/objects');

function generateMongoFixture(base, overrides) {
  return deepFreeze({
    ...base,
    _id: new mongoose.Types.ObjectId().toString(),
    ...overrides,
  });
}

async function resetDb() {
  const removals = Object.keys(mongoose.connection.collections).map(
    collection =>
      new Promise(resolve =>
        mongoose.connection.collections[collection].remove(resolve)
      )
  );

  let dropIndices = [];
  // dropIndices = Object.keys(mongoose.connection.collections).map(
  //   collection =>
  //     new Promise(resolve =>
  //       mongoose.connection.collections[collection].dropAllIndexes(resolve)
  //     )
  // );

  await Promise.all(_.flatten([removals, dropIndices]));
}

function setFixtureLocation(fixture, coordinates) {
  return {
    ...fixture,
    location: {
      ...fixture.location,
      coordinates: {
        type: 'Point',
        coordinates: coordinates,
      },
    },
  };
}

module.exports = {
  resetDb,
  setFixtureLocation,
  generateMongoFixture,
};
