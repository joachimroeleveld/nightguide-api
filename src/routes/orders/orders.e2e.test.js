require('../../shared/__test__/testBootstrap');

const update = require('immutability-helper');
const request = require('supertest');
const sinon = require('sinon');

const { validator } = require('../../shared/openapi');
const {
  TEST_ORDER_1,
  TEST_ORDER_2,
  TEST_ORDER_3,
} = require('../../shared/__test__/fixtures');
const { resetDb } = require('../../shared/__test__/testUtils');
const orderRepository = require('./orderRepository');
const { ORDER_STATUSES } = require('../../shared/constants');

const ORDER_SNAPSHOT_MATCHER = {
  id: expect.any(String),
  createdAt: expect.any(String),
  updatedAt: expect.any(String),
  items: expect.any(Array),
};

const ORDER_ITEM_SNAPSHOT_MATCHER = {
  id: expect.any(String),
};

const sandbox = sinon.createSandbox();

describe('orders e2e', () => {
  beforeEach(async () => {
    sandbox.restore();
    await resetDb();
  });

  describe('GET /orders', () => {
    const validateResponse = validator.validateResponse('get', '/orders');

    it('happy path', async () => {
      await orderRepository.createOrder(TEST_ORDER_1);

      const res = await request(global.app).get('/orders');

      expect(res.status).toEqual(200);
      expect(res.body.totalCount).toEqual(1);
      expect(res.body.results[0]).toMatchInlineSnapshot(
        ORDER_SNAPSHOT_MATCHER,
        `
Object {
  "__v": 0,
  "billingDetails": Object {
    "address": Object {
      "country": "NL",
    },
    "email": "foo@bar.nl",
    "name": "Foo Bar",
  },
  "createdAt": Any<String>,
  "id": Any<String>,
  "items": Any<Array>,
  "paymentMethod": Object {
    "type": "ideal",
  },
  "status": "completed",
  "updatedAt": Any<String>,
}
`
      );
      expect(res.body.results[0].items[0]).toMatchInlineSnapshot(
        ORDER_ITEM_SNAPSHOT_MATCHER,
        `
Object {
  "currency": "EUR",
  "id": Any<String>,
  "price": 10,
  "quantity": 1,
  "sku": "abc",
}
`
      );
      expect(validateResponse(res)).toBeUndefined();
    });

    it('should limit the amount of results to the limit parameter', async () => {
      const order1 = await orderRepository.createOrder(TEST_ORDER_1);
      await orderRepository.createOrder(TEST_ORDER_2);

      const res = await request(global.app)
        .get('/orders')
        .query({
          limit: 1,
        });

      expect(res.status).toEqual(200);
      expect(res.body.results.length).toEqual(1);
      expect(res.body.limit).toEqual(1);
      expect(res.body.results[0].id).toEqual(order1._id.toString());
      expect(validateResponse(res)).toBeUndefined();
    });

    it('should skip items set in offset parameter', async () => {
      await orderRepository.createOrder(TEST_ORDER_1);
      const order2 = await orderRepository.createOrder(TEST_ORDER_2);

      const res = await request(global.app)
        .get('/orders')
        .query({
          offset: 1,
        });

      expect(res.status).toEqual(200);
      expect(res.body.results.length).toEqual(1);
      expect(res.body.offset).toEqual(1);
      expect(res.body.results[0].id).toEqual(order2._id.toString());
      expect(validateResponse(res)).toBeUndefined();
    });

    it('returns a totalcount with each result', async () => {
      await orderRepository.createOrder(TEST_ORDER_1);
      await orderRepository.createOrder(TEST_ORDER_2);

      const res = await request(global.app)
        .get('/orders')
        .query({ limit: 1 });

      expect(res.status).toEqual(200);
      expect(res.body.results.length).toBe(1);
      expect(res.body.totalCount).toBe(2);
      expect(validateResponse(res)).toBeUndefined();
    });

    it('ids filter', async () => {
      const order1 = await orderRepository.createOrder(TEST_ORDER_1);
      const order2 = await orderRepository.createOrder(TEST_ORDER_2);
      await orderRepository.createOrder(TEST_ORDER_3);

      const ids = [order1._id.toString(), order2._id.toString()].sort();

      const res = await request(global.app)
        .get('/orders')
        .query({
          ids,
        });

      expect(res.status).toEqual(200);
      expect(res.body.results.length).toBe(2);
      expect(res.body.results.map(item => item.id).sort()).toEqual(ids);
      expect(validateResponse(res)).toBeUndefined();
    });
  });

  describe('GET /orders/:orderId', () => {
    const validateResponse = validator.validateResponse(
      'get',
      '/orders/{orderId}'
    );

    it('happy path', async () => {
      const order1 = await orderRepository.createOrder(TEST_ORDER_1);

      const res = await request(global.app)
        .get(`/orders/${order1._id.toString()}`)
        .send()
        .expect(200);

      expect(res.body.id).toEqual(order1._id.toString());
      expect(res.body).toMatchSnapshot(ORDER_SNAPSHOT_MATCHER);
      expect(validateResponse(res)).toBeUndefined();
    });

    it('simple fields', async () => {
      const order1 = await orderRepository.createOrder(
        update(TEST_ORDER_1, {
          billingDetails: {
            address: {
              city: { $set: 'Magalhaesstraat 6' },
              line1: { $set: 'Vechtplantsoen 56' },
              line2: { $set: '1' },
              postalCode: { $set: '3554 TG' },
              state: { $set: 'California' },
            },
          },
          items: {
            [0]: {
              eventId: { $set: '5d95ed62aee3f4f9c18c04e5' },
            },
          },
          paymentMethod: {
            ideal: {
              $set: {
                bank: 'ing',
              },
            },
          },
        })
      );

      const res = await request(global.app)
        .get(`/orders/${order1._id}`)
        .send()
        .expect(200);

      expect(res.body).toMatchSnapshot(ORDER_SNAPSHOT_MATCHER);
      expect(validateResponse(res)).toBeUndefined();
    });
  });

  describe('POST /orders', () => {
    const validateResponse = validator.validateResponse('post', '/orders');

    it('happy path', async () => {
      const res = await request(global.app)
        .post('/orders')
        .send(orderRepository.deserialize(TEST_ORDER_1));
      const body = res.body;

      expect(res.status).toEqual(201);
      expect(body.id).toEqual(TEST_ORDER_1._id);
      expect(validateResponse(res)).toBeUndefined();
    });
  });

  describe('PUT /orders', () => {
    const validateResponse = validator.validateResponse(
      'put',
      '/orders/{orderId}'
    );

    it('happy path', async () => {
      const order1 = await orderRepository.createOrder(TEST_ORDER_1);

      const res = await request(global.app)
        .put(`/orders/${order1._id}`)
        .send({
          status: ORDER_STATUSES.ORDER_STATUS_PENDING,
        });
      const body = res.body;

      expect(res.status).toEqual(200);
      expect(body.id).toEqual(TEST_ORDER_1._id);
      expect(body.status).toEqual(ORDER_STATUSES.ORDER_STATUS_PENDING);
      expect(validateResponse(res)).toBeUndefined();
    });
  });

  describe('DELETE /orders/:orderId', () => {
    const validateResponse = validator.validateResponse(
      'delete',
      '/orders/{orderId}'
    );

    it('happy path', async () => {
      const createdConfig = await orderRepository.createOrder(TEST_ORDER_1);

      const res = await request(global.app).delete(
        `/orders/${createdConfig._id}`
      );

      let order = await orderRepository.getOrder(createdConfig._id);

      expect(order).toBe(null);
      expect(res.status).toEqual(200);
      expect(validateResponse(res)).toBeUndefined();
    });
  });
});
