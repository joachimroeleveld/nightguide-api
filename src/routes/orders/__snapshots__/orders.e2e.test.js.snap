// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`orders e2e GET /orders/:orderId happy path 1`] = `
Object {
  "__v": 0,
  "billingDetails": Object {
    "address": Object {
      "country": "NL",
    },
    "email": "foo@bar.nl",
    "name": "Foo Bar",
  },
  "createdAt": Any<String>,
  "id": Any<String>,
  "items": Any<Array>,
  "paymentMethod": Object {
    "type": "ideal",
  },
  "status": "completed",
  "updatedAt": Any<String>,
}
`;

exports[`orders e2e GET /orders/:orderId simple fields 1`] = `
Object {
  "__v": 0,
  "billingDetails": Object {
    "address": Object {
      "city": "Magalhaesstraat 6",
      "country": "NL",
      "line1": "Vechtplantsoen 56",
      "line2": "1",
      "postalCode": "3554 TG",
      "state": "California",
    },
    "email": "foo@bar.nl",
    "name": "Foo Bar",
  },
  "createdAt": Any<String>,
  "id": Any<String>,
  "items": Any<Array>,
  "paymentMethod": Object {
    "ideal": Object {
      "bank": "ing",
    },
    "type": "ideal",
  },
  "status": "completed",
  "updatedAt": Any<String>,
}
`;
